client_plugin/drivers/shared/kvstore/etcdops/etcdops.go

```diff
@@ -332,6 +335,63 @@ func (e *EtcdKVS) etcdWatcher(cli *etcdClient.Client) {
 	}
 }
 
+// serviceAndVolumeGC: garbage collector for orphan services or volumes
+func (e *EtcdKVS) serviceAndVolumeGC(cli *etcdClient.Client) {
+	ticker := time.NewTicker(gcTicker)
+	quit := make(chan struct{})
+
+	for {
+		select {
+		case <-ticker.C:
+			// find all the vShared volume services
+			volumesToVerify, err := e.dockerOps.ListVolumesFromServices()
+			if err != nil {
+				log.Warningf("Failed to get vShared volumes according to docker services")
+			} else {
+				e.cleanOrphanServiceAndVolume(volumesToVerify, true)
+			}
+
+			// find all the internal volumes for vShared volume
+			volumesToVerify, err = e.dockerOps.ListVolumesFromInternalVol()
+			if err != nil {
+				log.Warningf("Failed to get internal volumes from docker")
+			} else {
+				e.cleanOrphanServiceAndVolume(volumesToVerify, false)
+			}
+		case <-quit:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// trimVolName: trim the volume name if there is special split characters existing in the name
+func trimVolName(volName string) string {
+	// Currently we only take @ as the split character
+	// TODO: need to take care of volumes with same name but on different datastores
+	s := strings.Split(volName, "@")
+	return s[0]
+}
+
+// cleanOrphanServiceAndVolume: stop orphan services and delete orphan internal volumes
+func (e *EtcdKVS) cleanOrphanServiceAndVolume(volumesToVerify []string, stopService bool) {
```

Comment:
Volume with "Error" state will stay until the user executes "docker volume rm" on this shared volume. This function will only remove service and internal volumes if the corresponding shared volume metadata is not in etcd, or the state of shared volume is "Deleting" or "Error". It won't delete shared volume metadata here.